Polished up the Oracle contract

JasperSlimmens opened the pull request:
I ran through the Oracle contract to see what I could optimize. Changed some logic and simplified some things. Deployment gas down from 5.40m gas to 5.04m gas.

belbix commented on contracts/OracleMainnet.sol:
```diff
@@ -251,33 +245,28 @@ contract OracleMainnet is Governable {
+        priceToken = computePrice(tokens[i]);
+        if (priceToken == 0) {
+          price = 0;
+          return price;
```
return 0; ?

JasperSlimmens replied:
This for an LP token, so underlying is multiple tokens. If one of these tokens does not return a price (so priceToken == 0). I return price = 0 for the LP token. Else the value will be inaccurate, so better to return 0.

belbix commented on contracts/OracleMainnet.sol:
```diff
@@ -251,33 +245,28 @@ contract OracleMainnet is Governable {
       return price;
     } else {
-      uint256 price = computePrice(token);
+      price = computePrice(token);
```
looks like we don't need the price variable

JasperSlimmens replied:
We do, it is returned on the next line. This is price computation for when input is a single token.

JasperSlimmens replied:
Get what you mean now. Put `return computePrice(token);`

belbix commented on contracts/OracleMainnet.sol:
```diff
@@ -288,47 +277,34 @@ contract OracleMainnet is Governable {
-        return true;
-      } else {
+    } else if (keccak256(abi.encodePacked(symbol)) == keccak256(abi.encodePacked(sushiSymbol))) {
+      try pair.factory{gas:3000}() returns (address factory) {
```
duplicate code
looks like we can move this code to a function

JasperSlimmens replied:
So I did this, but it is more expensive gas to deploy.

![image](https://user-images.githubusercontent.com/34768229/109773391-6f5a5500-7bff-11eb-9f07-4feec5267744.png)

JasperSlimmens replied:
```
//Checks if address is Uni or Sushi LP. This is done in two steps, because the second step seems to cause errors for some tokens.
  //Only the first step is not deemed accurate enough, as any token could be called UNI-V2.
  function isUniSushiCheck(address token) public view returns (bool) {
    IUniswapV2Pair pair = IUniswapV2Pair(token);
    string memory uniSymbol = "UNI-V2";
    string memory sushiSymbol = "SLP";
    string memory symbol = pair.symbol();
    bool checkUni = symbolLPCheck(pair, symbol, uniSymbol, uniswapFactoryAddress);
    bool checkSushi = symbolLPCheck(pair, symbol, sushiSymbol, sushiswapFactoryAddress);
    return (checkUni || checkSushi);
  }

  function symbolLPCheck(IUniswapV2Pair pair, string memory symbol, string memory compareSymbol, address compareFactory) public view returns (bool) {
    if (keccak256(abi.encodePacked(symbol)) == keccak256(abi.encodePacked(compareSymbol))) {
      try pair.factory{gas: 3000}() returns (address factory) {
        if (factory == compareFactory){
          return true;
        } else {
          return false;
        }
      } catch {
        return false;
      }
    } else {
      return false;
    }
  }
```

belbix replied:
code quality and readability have higher priority

belbix commented on contracts/OracleMainnet.sol:
```diff
@@ -288,47 +277,34 @@ contract OracleMainnet is Governable {
-    try pair.symbol() returns (string memory symbol) {
-      if (keccak256(abi.encodePacked(symbol)) != keccak256(abi.encodePacked(sushiSymbol))) {
+    string memory symbol = pair.symbol();
+    if (keccak256(abi.encodePacked(symbol)) == keccak256(abi.encodePacked(uniSymbol))) {
```
maybe better create `is***` function with arguments for this?

JasperSlimmens replied:
I am not sure what you mean

belbix replied:
a function with two args that compare values
isEqualSymbol (I guess not the best name)
keccak256(abi.encodePacked(symbol)) == keccak256(abi.encodePacked(uniSymbol))

belbix replied:
maybe just `isEqual`

JasperSlimmens replied:
![image](https://user-images.githubusercontent.com/34768229/109796862-f831ba80-7c18-11eb-89d6-a7a39a6c9614.png)

belbix commented on contracts/OracleMainnet.sol:
```diff
@@ -468,46 +438,49 @@ contract OracleMainnet is Governable {
   }
 
   //Gives the Uniswap pool with largest liquidity for a given token and a given tokenset (either keyTokens or pricingTokens)
   function getUniLargestPool(address token, address[] memory tokenList) public view returns (address, uint256) {
```
could you union this function with Sushi checking? just add necessary vars to arguments

JasperSlimmens replied:
Yes, that a good idea. Will do.

JasperSlimmens replied:
![image](https://user-images.githubusercontent.com/34768229/109803012-d3414580-7c20-11eb-83f4-227d92cec711.png)

JasperSlimmens commented:
Updated the contract with all comments. 

- Cleaned up `isUniSushiCheck()`. 
- Merged the `getUniLargestPool()` and `getSushiLargestPool()` functions. 
- Had a look at function visibility. Changed many functions to internal. Internal functions are much cheaper, but might be nice to have at least some of the functions available for external callers. If we wanted we could make UI ultimately simple by only having the `getPrice()` function available.
